---
<!--
  %% \VignetteEngine{knitr::knitr}
  -->
title: "Easily run a RNA-Seq analysis via RNASeqHelper"
shortitle: "RNASeqHelper Vignette"
date: "`r Sys.Date()`"
author:
- name: "Mehmet Tekman"
  affiliation: Institute of Experimental and Clinical Pharmacology and Toxicology, University Freiburg
  email: mtekman89@gmail.com
- name: "Sebastian Arnold"
  affiliation: Institute of Experimental and Clinical Pharmacology and Toxicology, University Freiburg
  email: sebastian.arnold@pharmakol.uni-freiburg.de
package: RNASeqHelper
abstract: >
    Our work introduces an R package for interactive RNA-Seq analysis, enhancing DESeq2's robust statistical methods. It simplifies complex analyses, offering user-friendly features like heatmap and volcano plot generation. This package serves as both an interactive tool and an automated pipeline, enabling efficient exploration of RNA-Seq data without extensive programming knowledge. Its intuitive visualization and seamless integration with DESeq2 facilitate gene expression studies, fostering discoveries.
output: 
  BiocStyle::pdf_document
vignette: >
    %\VignetteIndexEntry{Easily run a RNA-Seq analysis via RNASeqHelper}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---
\tableofcontents
```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

# RNASeqHelper

This vignette walks through an RNA-seq analysis using simulated datasets.

Let us first load the library

```{r}
library(RNASeqHelper)
```



## Input Matrices and Annotation

Every RNA-seq analysis begins with two datasets:

- The input matrix
- The phenotype data


#### Simulate A Dataset

Here we create 9 RNA samples with slighly varying means and SDs, for
10000 genes.

```{r}
num_genes <- 10000
mean_sd <- list(c(10,10), c(10,10), c(10,10),
                c(20,20), c(20,20), c(30,30),
                c(90,90), c(90,90), c(90,90))

mat <- as.matrix(
    do.call(cbind,
        lapply(seq_along(mean_sd), function(i){
    x = mean_sd[[i]]
    tab <- matrix(as.integer(rnorm(num_genes, x[[1]], x[[2]])), ncol=1,
                dimnames=list(paste0("G", 1:num_genes), paste0("S", i)))
    tab = tab - min(tab) ## no zeroes
    return(tab)
})))

phenotype_data = data.frame(sample = colnames(mat),
            condition = c(rep("green", 3),rep("red", 3), rep("blue", 3)),
            time = c(5, 5, 15, 10, 10, 15, 5, 10, 15))

message("First 6 lines of the matrix")
head(mat)
message("Phenotype Data")
phenotype_data
```

So now we have a matrix of 9 samples and 10000 genes, and accompanying
phenotype data for the samples.


#### Genes of Interest

We also might want some gene plots for some genes of interest. We can
either load some pre-defined lists via:

```{r}
gois_list <- source(system.file("extdata", "gois.R", package="RNASeqHelper"))
gois_list
```

Or we can make our own. Note that the genes above are good defaults
for real (Mouse) datasets, but will not work in the remainder of this
vignette.

We will instead construct our own, following the schema of "name" ->
vector of genes.

```{r}
gois_list <- list(apple = paste0("G", 100:105),
                banana = paste0("G", 810:819),
                carrot = paste0("G", 256:260))
gois_list
```

## RNA-Seq Analysis (Exploratory)

Let us now perform an analysis.

It is possible to do the entire analysis with a single call to the
`rnaseqhelper` function, but it's sometimes better to perform a more
exploratory analysis.


#### Selecting for Informative Genes

Not all of our genes will be useful for RNA-seq analysis, so we need
to select for an informative subset using two parameters:

- `min_detect`, the minimum threshold of expression a gene must
    express in a sample before it can count as "detected".
- `min_occur`, the minimum occurences of a "detected" gene across all
    samples.

Here we will use 3 and 2, selecting for genes that occur in at least 3
samples with a minumum expression of 2.

```{r}
keep_genes <- high_quality_genes(mat, min_occur=3, min_detect=2,
                                out_dir=file.path(tempdir(), "0_genes"))
```

We are told that no genes were dropped.

We can inspect this ourselves by examining the output file
"0_genes/smallestGroup3-detected2-keep_genes".

Usually we expect a few genes to be dropped, so let's increase the
parameters:

```{r}
keep_genes <- high_quality_genes(mat, min_occur=8, min_detect=20,
                                out_dir=file.path(tempdir(), "0_genes"))
```

Here we see 87 genes are dropped, leaving us with 92% of the data
still, and we can once again inspect this directly by looking at the
"0_genes/smallestGroup9-detected10-keep_genes" file


#### Validating for Informative Genes

We will now perform a preliminary DESeq2 analysis using our filtered
genes and using the `phenotype_data` to select for sample groups.

The `condition` column in the `phenotype_data` is used to group are
samples, and is an implicit declaration.

```{r}
res <- run_deseq(mat, keep_genes, phenotype_data,
        out_dir = file.path(tempdir(), "1_matrices_and_deseq"))
```

Here tables and plots were written to the "1_matrices_and_deseq" folder:

- `deseq2obj.rds`, an RData object of the DESeq2 object,
    useful mostly for debugging.
- `phenotype_data.tsv`, the input phenotype data table.
- `input_matrix.tsv`, the input matrix table.
- `input_matrix.filt.tsv`, the above matrix, filtered for the genes kept.
- `input_matrix.filt.normalized.tsv`, the above filtered matrix,
    normalized by DESeq2.
- `input_matrix.filt.normalized.vst_corrected.tsv`, the filtered matrix with
    a Variance Stabilizing Transform to correct the data.
- `input_matrix.filt.normalized.vst_corrected_PCA.pdf`, a PCA of the
    normalized and transformed values. 

What you should be seeing at this stage are samples of the same type
clustered together. The transformed values might cluster better than
the normalized values, so it is up to you which one of these you
choose for downstream analysis.


#### Heatmaps and Volcano Plots

We will now perform the rest of the analysis, by searching for
differentially expressed (DE) genes via contrasting sample groups
against one another, and producing pairwise (sample group1 vs sample
group2) kmeans heatmaps and global (all samples) kmeans heatmaps for a
variety of k-values.

Let's contrast our green and blue samples against one another.

We will be using the output of the `run_deseq` function in the last
step which we called `res`. We will also use our genes of interest.

```{r}
pairwise_hmap_volcano(res$ddsObj,
                    transformed_counts = assay(res$vFalse),
                    numer = "green", denom = "blue",
                    top_ngenes_tocluster = 2000,
                    top_ngenes_tohighlight = 25,
                    score_thresh = c(0.5, 0.9),
                    genes_of_interest = gois_list,
                    kmeans_list = c(2,5),
                    out_dirprefix=file.path(tempdir(), "2_heatmaps"))

```

The volcano plots are generated from the contrasts of our two sample
groups "green" and "blue". The DE genes from these contrasts are then
used to cluster the samples pairwise and globally for different Kmeans
k-values.

The results of all can be found in the "~/myanalysis/2_heatmaps" folder.

## RNA-Seq Analysis (Pipeline)

It is entirely possible to run the complete analysis above with a
single function. Here we change some parameters to speed up the
analysis.


```{r}
rnaseqhelper(mat, phenotype_data,
            out_dir=file.path(tempdir(), "second_analysis"),
            numer = "green", denom = "blue",
            keep_params = list(min_occur=8, min_detect=20),
            heat_params = list(top_ngenes_tocluster = 1000,
                                top_ngenes_tohighlight = 25,
                                score_thresh = c(0.9),
                                genes_of_interest = gois_list["apple"],
                                kmeans_list = c(2)))

```


## SessionInfo

```{r}
sessionInfo()
```
