---
title: "RNASeqHelper reference manual"
date: "`r Sys.Date()`"
author: "Mehmet Tekman"
output:
  html_document:
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{Example CustomGenome analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
\tableofcontents
```{r echo=FALSE}
knitr::opts_chunk$set(fig.width=8, fig.height=8, dpi=50, dev='jpeg')
```

# RNASeqHelper

This vignette enables an RNA-seq analysis


## Input Matrices and Annotation



## Analysis


```R
tab = as.matrix(do.call(cbind, 
                        lapply(list.files("3_counts/", pattern="*.tab",
                               full.names=T), 
                        function(x){ read.table(x,row.names=1,
                        check.names=FALSE) })))
colnames(tab) = gsub(
    "\\.TAF2_dIDR", "",
    gsub("\\.align.*", "", list.files("3_counts/", pattern="*.tab"))
)                 

phenotype_data =  
    data.frame(Sample = colnames(tab),
               Condition = gsub("_\\d$", "", colnames(tab)),
               Replicate = gsub("^.*_", "", colnames(tab)))

keep_genes = high_quality_genes(tab, min_occur=3, min_detect=2, 
                                out_dir="test/1_genes")

res = run_deseq(tab, keep_genes, phenotype_data)

pca_and_matrices(res, out_dir="test/input_matrices_and_deseq")

pairwise_hmap_volcano(res$ddsObj,
                      transformed_counts = assay(res$vFalse),
                      numer = "NLS", denom = "TAF2",
                      top_ngenes_tocluster = 2000,
                      top_ngenes_tohighlight = 25,
                      lFC_zoom = 1.5, pAdj_zoom = 20,
                      kmeans = c(2,5,8),
                      out_dirprefix="test/outputs")

## Todo gene plots
```

## SessionInfo

```R
sessionInfo()
```
